### DISCUSSION TITLE
errors while updating Zowe SDK create session function to Zowe V2 version

### DISCUSSION THREAD

**Discussion by ghost** [zowe/zowe-cli#1548] [Q&A] [answered]
I am facing the same issue in fixing this, if anyone can able to help, pls do. 
Below code snippet where i am trying to replace the with new api version of zowe v2. But i am into getting error. 
 Could anyone pls compare the old code with new code and help me, what is the mistake i make here pls?

i referred this case, while fixing it. 
link : https://github.com/zowe/zowe-cli/issues/1141

Old code:

const zowe = require('@zowe/cli')
const zowejobs = require('@zowe/zos-jobs-for-zowe-sdk')
const IssueTso = require("@zowe/zos-tso-for-zowe-sdk").IssueTso;

let userTSO = process.env.MF_USER_TSO;
let paramsTSO = {};
const host = process.env.MF_HOST;
const port = process.env.MF_PORT;
const user = process.env.MF_USER_ZOS;
const password = process.env.MF_PASSWORD;
const profile = {
    host, port, user, password, rejectUnauthorized: false
}

router.get('/getbulletin', async (req, res) => {
    console.log('Inside MF router get /getbulletin');
    console.log("req.body: ", req.body);
    
    let command = `ex 'z44168.test.clist(soops)' 'BULLETIN'`;
    try {
        **const session = zowe.ZosmfSession.createBasicZosmfSessionFromArguments(profile)**
        let response = await IssueTso.issueTsoCommand(session, userTSO, command, paramsTSO);
        console.log("original response --------------------------------------------------");
        console.log(response)
        console.log("original response --------------------------------------------------");
        if (response && response.success) {
            let responseStr = response.commandResponse;
            console.log("responseStr --------------------------------------------------");
            console.log(responseStr);
            console.log("responseStr --------------------------------------------------");
            if (!responseStr || responseStr.length == 0) {
                res.status(201).json({message: 'Invalid or No response from Mainframe'});
            }
            responseStr = responseStr.replace(/READY/g, '');
            responseStr = responseStr.replace(/\n/g, '');
            responseStr = responseStr.replace(/</g, '[');
            responseStr = responseStr.replace(/>/g, ']');
            // responseStr = responseStr.replace(/ /g, '');
            console.log("formatted responseStr --------------------------------------------------");
            console.log(responseStr);
            console.log("formatted responseStr --------------------------------------------------");
            try {
                JSON.parse(responseStr);
                res.status(201).json(JSON.parse(responseStr));
            } catch (error) {
                console.log(error);
                res.status(201).json({message: 'Invalid JSON format from Mainframe'});
            }
        } else {
            res.status(201).json({message: 'Invalid or No response from Mainframe'});
        }
    } catch (err) {
        console.log(err)
        res.status(500).json(err);
    }
});

New code:

const zowe = require('@zowe/cli')
let userTSO = process.env.MF_USER_TSO;
let paramsTSO = {};
const host = process.env.MF_HOST;
const port = process.env.MF_PORT;
const user = process.env.MF_USER_ZOS;
const password = process.env.MF_PASSWORD;
const profile = {
    host, port, user, password, rejectUnauthorized: false
}

router.get('/getbulletin', async (req, res) => {
    console.log('Inside MF router get /getbulletin');
    console.log("req.body: ", req.body);
    
    let command = `ex 'z44168.test.clist(soops)' 'BULLETIN'`;
    try {
        const session = zowe.ZosmfSession.createSessCfgFromArgs(profile)
        let response = await IssueTso.issueTsoCommand(session, userTSO, command, paramsTSO);
        console.log("original response --------------------------------------------------");
        console.log(response)
        console.log("original response --------------------------------------------------");
        if (response && response.success) {
            let responseStr = response.commandResponse;
            console.log("responseStr --------------------------------------------------");
            console.log(responseStr);
            console.log("responseStr --------------------------------------------------");
            if (!responseStr || responseStr.length == 0) {
                res.status(201).json({message: 'Invalid or No response from Mainframe'});
            }
            responseStr = responseStr.replace(/READY/g, '');
            responseStr = responseStr.replace(/\n/g, '');
            responseStr = responseStr.replace(/</g, '[');
            responseStr = responseStr.replace(/>/g, ']');
            // responseStr = responseStr.replace(/ /g, '');
            console.log("formatted responseStr --------------------------------------------------");
            console.log(responseStr);
            console.log("formatted responseStr --------------------------------------------------");
            try {
                JSON.parse(responseStr);
                res.status(201).json(JSON.parse(responseStr));
            } catch (error) {
                console.log(error);
                res.status(201).json({message: 'Invalid JSON format from Mainframe'});
            }
        } else {
            res.status(201).json({message: 'Invalid or No response from Mainframe'});
        }
    } catch (err) {
        console.log(err)
        res.status(500).json(err);
    }
});

Error received :
Inside MF router get /getbulletin
req.body:  {}
Callinggg --------------------------------------------------
after Callinggg --------------------------------------------------[object Object]
TypeError: Cannot read property 'hostname' of undefined
    at ZosmfRestClient.buildOptions (C:\Rajesh Files\personal\GANESH FILES\prod\server_v3\node_modules\@zowe\imperative\lib\rest\src\client\AbstractRestClient.js:261:45)
    at Promise (C:\Rajesh Files\personal\GANESH FILES\prod\server_v3\node_modules\@zowe\imperative\lib\rest\src\client\AbstractRestClient.js:117:39)
    at new Promise (<anonymous>)
    at ZosmfRestClient.request (C:\Rajesh Files\personal\GANESH FILES\prod\server_v3\node_modules\@zowe\imperative\lib\rest\src\client\AbstractRestClient.js:95:16)
    at Function.postExpectString (C:\Rajesh Files\personal\GANESH FILES\prod\server_v3\node_modules\@zowe\imperative\lib\rest\src\client\RestClient.js:226:34)
    at Function.<anonymous> (C:\Rajesh Files\personal\GANESH FILES\prod\server_v3\node_modules\@zowe\imperative\lib\rest\src\client\RestClient.js:88:37)
    at Generator.next (<anonymous>)
    at C:\Rajesh Files\personal\GANESH FILES\prod\server_v3\node_modules\@zowe\imperative\lib\rest\src\client\RestClient.js:18:71   
    at new Promise (<anonymous>)
    at __awaiter (C:\Rajesh Files\personal\GANESH FILES\prod\server_v3\node_modules\@zowe\imperative\lib\rest\src\client\RestClient.js:14:12)
GET /mf/getbulletin 500 31.050 ms - 2
Inside MF router get /mf/getjoblist
TypeError: Cannot read property 'hostname' of undefined
    at ZosmfRestClient.buildOptions (C:\Rajesh Files\personal\GANESH FILES\prod\server_v3\node_modules\@zowe\imperative\lib\rest\src\client\AbstractRestClient.js:261:45)
    at Promise (C:\Rajesh Files\personal\GANESH FILES\prod\server_v3\node_modules\@zowe\imperative\lib\rest\src\client\AbstractRestClient.js:117:39)
    at new Promise (<anonymous>)
    at ZosmfRestClient.request (C:\Rajesh Files\personal\GANESH FILES\prod\server_v3\node_modules\@zowe\imperative\lib\rest\src\client\AbstractRestClient.js:95:16)
    at Function.getExpectString (C:\Rajesh Files\personal\GANESH FILES\prod\server_v3\node_modules\@zowe\imperative\lib\rest\src\client\RestClient.js:196:34)
    at Function.<anonymous> (C:\Rajesh Files\personal\GANESH FILES\prod\server_v3\node_modules\@zowe\imperative\lib\rest\src\client\RestClient.js:50:37)
    at Generator.next (<anonymous>)
    at C:\Rajesh Files\personal\GANESH FILES\prod\server_v3\node_modules\@zowe\imperative\lib\rest\src\client\RestClient.js:18:71   
    at new Promise (<anonymous>)
    at __awaiter (C:\Rajesh Files\personal\GANESH FILES\prod\server_v3\node_modules\@zowe\imperative\lib\rest\src\client\RestClient.js:14:12)
GET /mf/getjoblist 500 42.440 ms - 32

**Comment by t1m0thyj** [comment, marked as answer]
@inv9wj1 Here are two methods that should work to convert the `profile` object into a session class:
1. The `addPropsOrPrompt` method must be awaited. It will issue a CLI prompt for host/user/password if they are missing.
```
const profile = {
    host: "example.com", port: 443, user: "admin", password: "123456", rejectUnauthorized: false
};
const sessCfg = zowe.ZosmfSession.createSessCfgFromArgs(profile);
const sessCfgWithCreds = await imperative.ConnectionPropsForSessCfg.addPropsOrPrompt(sessCfg, profile);
const session = new imperative.Session(sessCfgWithCreds);
```
2. If you don't want prompting, you can convert the profile to an array of args and pass it to the `createSession` method.
```
const profile = {
    host: "example.com", port: 443, user: "admin", password: "123456", rejectUnauthorized: false
};
const profArgs = Object.entries(profile).map(([k, v]) => ({ argName: k, argValue: v }));
const session = imperative.ProfileInfo.createSession(profArgs);
```

**Reply by ghost** [reply]
Thanks Much for looking into it, i tried the suggested solution, i could see its trying to create a session, but getting different error this time.  though the host and port is correct, it looks for something else. Could you help me if you have any idea on this error pls?


{ msg: 
      'z/OSMF REST API Error:\nhttp(s) request error event called\nError: self signed certificate in certificate chain',
     causeErrors:
      { Error: self signed certificate in certificate chain
          at TLSSocket.onConnectSecure (_tls_wrap.js:1088:34)
          at TLSSocket.emit (events.js:198:13)
          at TLSSocket._finishInit (_tls_wrap.js:666:8) code: 'SELF_SIGNED_CERT_IN_CHAIN' },
     source: 'client',
     errorCode: undefined,
     protocol: 'https',
     port: '10443',
     host: '192.86.32.67',
     basePath: '',
     httpStatus: undefined,
     errno: undefined,
     syscall: undefined,
     payload: undefined,
     headers: [ [Object] ],
     resource: '/zosmf/restjobs/jobs?owner=z44168',
     request: 'GET',
     additionalDetails:
      'HTTP(S) client encountered an error. Request could not be initiated to host.\nReview connection details (host, port) and ensure correctness.\n\nProtocol: https\nHost: 192.86.32.67\nPort: 10443\nBase Path: \nResource: /zosmf/restjobs/jobs?owner=z44168\nRequest: GET\nHeaders: [{"X-CSRF-ZOSMF-HEADER":true}]\nPayload: undefined' },
  mMessage:
   'z/OSMF REST API Error:\nhttp(s) request error event called\nError: self signed certificate in certificate chain' }

**Reply by t1m0thyj** [reply]
This error means that the `rejectUnauthorized: false` is being ignored. I've updated the sample code above, could you please try the following:
```
const sessCfg = zowe.ZosmfSession.createSessCfgFromArgs(profile);
const sessCfgWithCreds = await imperative.ConnectionPropsForSessCfg.addPropsOrPrompt(sessCfg, profile);
```

The call to `createSessCfgFromArgs` should handle the rejectUnauthorized property. Also if you prefer to disable prompting, you could add an `{ doPrompting: false }` argument on the call to `addPropsOrPrompt`.

**Reply by ghost** [reply]
Thanks Much Timothy for the support, it works perfectly fine now.  :)
